### PR TITLE
[API] As a user, I can sign in with Google account

### PR DESCRIPTION
**Pull request by hanam1ni**
## What happened

- Setup `Ueberauth` to support API authentication endpoint 
- Setup `Guardian` to handle and manage access token
- Returns access token after user authenticated

## Proof Of Work

Access token is returned after success authentication

localhost_4000_api_auth_google_callback_code_4_0AY0e-g6Pv9-ht9AKiPJV-15aHApQDOgauOuaHPHfbMKCmDEsvLdbuFYe9PANm9NTaKZbVw_scope_email_https___www_googleapis_com_auth_userinfo_email_openid_authuser_0_hd_nimblehq_co_prompt_none

### CODE REVIEW

File: config/prod.exs [hunk 1]
```diff
@@ -55,3 +55,7 @@ config :logger, level: :info
 # Finally import the config/prod.secret.exs which loads secrets
 # and configuration from environment variables.
 import_config "prod.secret.exs"
+
+config :google_crawler, GoogleCrawler.Tokenizer,
+  issuer: "google_crawler",
+  secret_key: System.get_env("GUARDIAN_SECRET")
```
Comment (andyduong1920) [inline config/prod.exs:61]: how about moving this to `prod.secret.exs` @hanam1ni

File: lib/google_crawler/identities.ex [hunk 2]
```diff
@@ -0,0 +1,9 @@
+defmodule GoogleCrawler.Identities do
+  alias GoogleCrawler.Identities.User
+  alias GoogleCrawler.Repo
+
+  def get_user_by(params) do
+    User
+    |> Repo.get_by(params)
```
Comment (andyduong1920) [inline lib/google_crawler/identities.ex:7]: as the pipe has only 1 step, we could go without that @hanam1ni 

```
Repo.get_by(User, params)
```

File: config/prod.exs (hunk 1 above)
Comment (hanam1ni) [reply config/prod.exs:61]: Good idea 👍  In Omiselab, we don't use much of that file.

Comment (hanam1ni) [reply config/prod.exs]: Updated in this commit 710e3cba129dba0ff7c60bc9878040734bde869f

File: lib/google_crawler_web/controllers/api/auth_controller.ex [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        |> render("token.json", %{access_token: access_token})
+
+      {:error, _} ->
+        conn
+        |> ErrorHandler.handle(:bad_request)
+    end
+  end
+
+  defp insert_or_update_user(changeset) do
+    case Repo.get_by(User, email: changeset.changes.email) do
```
Comment (andyduong1920) [inline lib/google_crawler_web/controllers/api/auth_controller.ex:33]: a small thing, how about moving all of this logic into `GoogleCrawler.Identities.User` 🤔 so the controller no need to care about the Repo stuff....

File: test/google_crawler/identities_test.exs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+defmodule GoogleCrawler.IdentitiesTest do
+  use GoogleCrawler.DataCase
```
Comment (andyduong1920) [inline test/google_crawler/identities_test.exs:2]: maybe we could turn async true here ☁️

File: test/google_crawler/tokenizer_test.exs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+defmodule GoogleCrawler.TokenizerTest do
+  use GoogleCrawler.DataCase
```
Comment (andyduong1920) [inline test/google_crawler/tokenizer_test.exs:2]: how about adding `async: true` @hanam1ni

File: test/google_crawler_web/views/error_view_test.exs [hunk 6]
```diff
@@ -1,14 +1,18 @@
 defmodule GoogleCrawlerWeb.ErrorViewTest do
-  use GoogleCrawlerWeb.ConnCase, async: true
+  use GoogleCrawlerWeb.ViewCase
```
Comment (andyduong1920) [inline test/google_crawler_web/views/error_view_test.exs:2]: Any reason we need to turn off the `async: true` @hanam1ni 🤔

File: test/support/view_case.ex [hunk 7]
```diff
@@ -0,0 +1,12 @@
+defmodule GoogleCrawlerWeb.ViewCase do
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      use Mimic
```
Comment (andyduong1920) [inline test/support/view_case.ex:6]: I guess we don't need Mimic in the ViewCase yet

File: test/support/api_conn_case.ex [hunk 8]
```diff
@@ -0,0 +1,28 @@
+defmodule GoogleCrawlerWeb.ApiConnCase do
+  use ExUnit.CaseTemplate
+
+  using do
+    quote do
+      alias GoogleCrawlerWeb.ApiRouter.Helpers, as: Routes
+
+      use Mimic
```
Comment (andyduong1920) [inline test/support/api_conn_case.ex:8]: I guess we don't need Mimic yet

Comment (hanam1ni) [reply test/support/api_conn_case.ex:8]: To align with the `conn_case` and we will use Mimic in `keyword_controller_test`

**Review by andyduong1920 — APPROVED** [review]

File: test/google_crawler_web/views/api/auth_view_test.exs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+defmodule GoogleCrawlerWeb.Api.AuthViewTest do
+  use GoogleCrawlerWeb.ViewCase
```
Comment (andyduong1920) [inline test/google_crawler_web/views/api/auth_view_test.exs:2]: Just 1 small comment @hanam1ni I approved the PR but could you double-check all-new test files in this PR, can we add the `, async: true` to it? I commented on a few test files before but you could also check and add the `, async: true` to all test files

Comment (hanam1ni) [reply test/google_crawler_web/views/api/auth_view_test.exs:2]: @andyduong1920 Sure 👍  I will update in this PR

File: lib/google_crawler_web/controllers/api/auth_controller.ex [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  plug Ueberauth
+
+  alias GoogleCrawler.{Identities, Tokenizer}
+  alias GoogleCrawlerWeb.ErrorHandler
+
+  def callback(%{assigns: %{ueberauth_auth: auth}} = conn, _) do
+    user_params = %{token: auth.credentials.token, email: auth.info.email, provider: "google"}
+
+    signin(conn, user_params)
```
Comment (bterone) [inline lib/google_crawler_web/controllers/api/auth_controller.ex:12]: `signin` => `sign_in` ✍️

**Review by bterone — APPROVED** [review]

**Pull request merged**